packages/widget/src/hooks/useHeaderHeight.ts

```diff
@@ -1,25 +1,40 @@
-import { useHasExternalWalletProvider } from '../providers/WalletProvider/useHasExternalWalletProvider.js';
-import { useWidgetConfig } from '../providers/WidgetProvider/WidgetProvider.js';
-
-export const minHeaderHeight = 64;
-export const maxHeaderHeight = 108;
-export const maxHeaderHeightSubvariantSplit = 136;
+import { useLayoutEffect, useState } from 'react';
+import { useDefaultElementId } from '../hooks/useDefaultElementId.js';
+import { getHeaderElement } from '../utils/elements.js';
 
 // We use fixed position on the header when Widget is in Full Height layout.
 // We do this to get it to work like the sticky header does in the other layout modes.
 // As the header is position fixed its not in the document flow anymore.
 // To prevent the remaining page content from appearing behind the header we need to
 // pass the headers height so that the position of the page content can be adjusted
 export const useHeaderHeight = () => {
-  const { hiddenUI, subvariant } = useWidgetConfig();
-  const { hasExternalProvider } = useHasExternalWalletProvider();
+  const elementId = useDefaultElementId();
+  const headerElement = getHeaderElement(elementId);
+  const [headerHeight, setHeaderHeight] = useState(
+    headerElement?.getBoundingClientRect().height ?? 0,
+  );
+
+  useLayoutEffect(() => {
+    const handleHeaderResize = () => {
+      const height = headerElement?.getBoundingClientRect().height;
+
+      if (height) {
+        setHeaderHeight(height);
+      }
+    };
+
+    let resizeObserver: ResizeObserver;
+    if (headerElement) {
+      resizeObserver = new ResizeObserver(handleHeaderResize);
+      resizeObserver.observe(headerElement);
+    }
 
-  const headerHeight =
-    subvariant === 'split'
-      ? maxHeaderHeightSubvariantSplit
-      : hiddenUI?.includes('walletMenu') || hasExternalProvider
-        ? minHeaderHeight
-        : maxHeaderHeight;
+    return () => {
+      if (resizeObserver) {
+        resizeObserver.disconnect();
+      }
+    };
+  }, [elementId, headerHeight, headerElement, setHeaderHeight]);
```

Comment:
Using the MutationObserver here in place of the imperative if checks we were doing previously - that was getting more difficult to get to get it to work in certain situations.

For the split subvariant there was still this problem in full height layout mode. And this problem can happen on other pages.

![Screenshot 2024-09-24 at 19 42 46](https://github.com/user-attachments/assets/fa682f37-f21b-4201-98d2-e859d127fa67)

One improvement I think that would still be good here is to move the `useLayoutEffect` to the Header component itself and have that set a value on the Header Store - the useHeaderHeight hook would then become an access point for the height value in the store and using zustand shallow checks would actually be able to limit potential renders in the Widget where the height value it used

Reply:
@DNR500 are we sure we want to see `headerHeight` in the deps array?

Reply:
> One improvement I think that would still be good here is to move the useLayoutEffect to the Header component itself and have that set a value on the Header Store - the useHeaderHeight hook would then become an access point for the height value in the store and using zustand shallow checks would actually be able to limit potential renders in the Widget where the height value it used

Should we do it now? 👀

Reply:
I'm doing this move now and the deps array is much simpler and doesn't need the headerHeight.